test/gather_test.js

```diff
@@ -0,0 +1,283 @@
+      shape: [3],
+      data: [1, 0, 1],
+    };
+    const expected = {
+      shape: [3, 4, 3],
+      data: [
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+        0,   1,   2,
+        10,  11,  12,
+        20,  21,  22,
+        30,  31,  32,
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+      ],
+    };
+    testGather(input, indices, expected, {axis: 0});
+  });
+
+  it('gather 3D by 2D indices axis=0', function() {
+    const input = {
+      shape: [2, 4, 3],
+      data: [
+        0,   1,   2,
+        10,  11,  12,
+        20,  21,  22,
+        30,  31,  32,
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+      ],
+    };
+    const indices = {
+      shape: [3, 2],
+      data: [1, 0, 1, 0, 0, 1],
+    };
+    const expected = {
+      shape: [3, 2, 4, 3],
+      data: [
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+        0,   1,   2,
+        10,  11,  12,
+        20,  21,  22,
+        30,  31,  32,
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+        0,   1,   2,
+        10,  11,  12,
+        20,  21,  22,
+        30,  31,  32,
+        0,   1,   2,
+        10,  11,  12,
+        20,  21,  22,
+        30,  31,  32,
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+      ],
+    };
+    testGather(input, indices, expected, {axis: 0});
+  });
+
+  it('gather 3D by axis=1', function() {
+    const input = {
+      shape: [2, 4, 3],
+      data: [
+        0,   1,   2,
+        10,  11,  12,
+        20,  21,  22,
+        30,  31,  32,
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+      ],
+    };
+    const indices = {
+      shape: [2],
+      data: [2, 0],
+    };
+    const expected = {
+      shape: [2, 2, 3],
+      data: [
+        20,  21,  22,
+        0,   1,   2,
+        120, 121, 122,
+        100, 101, 102,
+      ],
+    };
+    testGather(input, indices, expected, {axis: 1});
+  });
+
+  it('gather 3D by 2D indices axis=2', function() {
+    const input = {
+      shape: [2, 4, 3],
+      data: [
+        0,   1,   2,
+        10,  11,  12,
+        20,  21,  22,
+        30,  31,  32,
+        100, 101, 102,
+        110, 111, 112,
+        120, 121, 122,
+        130, 131, 132,
+      ],
+    };
+    const indices = {
+      shape: [2, 2],
+      data: [0, 1, 1, 2],
+    };
+    const expected = {
+      shape: [2, 4, 2, 2],
+      data: [
+        0,   1,   1,   2,
+        10,  11,  11,  12,
+        20,  21,  21,  22,
+        30,  31,  31,  32,
+        100, 101, 101, 102,
+        110, 111, 111, 112,
+        120, 121, 121, 122,
+        130, 131, 131, 132,
+      ],
+    };
+    testGather(input, indices, expected, {axis: 2});
+  });
```

Comment:
We should include some 0D cases too. e.g.

```
    // 2D, axis 1, rank 0 indices tensor
    {
      "axis": 1,
      "data": [[1,2,3],
               [4,5,6],
               [7,8,9]],
      "indices": 1,
      "output": [2, 5, 8],
      "T": "float32"
    },
    // 0D, axis 0, rank 0 indices tensor
    {
      "axis": 0,
      "data": [1,2,3],
      "indices": 1,
      "output": 2,
      "T": "float32"
    },
```

Reply:
Which axis is for 0D scalar? I think axis 0 works for 1+ D tensor. The axis is in [0, N-1], for 0D, **N-1** = **-1** .

Reply:
I think that Dwayne means 0-D indices, which will cause the output rank less than the input rank.